command/auto_complete.go

```diff
@@ -0,0 +1,247 @@
+		cmd="${COMP_LINE:0:$COMP_POINT}"` +
+
+	// if we want to complete the second argurment and we didn't started writing
+	// yet then we should pass an empty string as another argument. Otherwise
+	// the white spaces will be discarded and the program will make suggestions
+	// as if it is completing the first argument.
+	// Beware that the we want to pass empty string so we intentionally write
+	// as it is. Fixes of SC2089 and SC2090 are not what we want.
+	// see also https://www.shellcheck.net/wiki/SC2090
+	`
+		[ "${COMP_LINE:COMP_POINT-1:$COMP_POINT}" == " " ] \
+			&& cmd="${cmd} \"\"" ` +
+
+	// execute the command with '--generate-bash-completion' flag to obtain
+	// possible completion values for current word.
+	// ps. SC2090 is not wanted.
+	`
+		opts=$($cmd --generate-bash-completion)` +
+
+	// prepare completion array with possible values and filter those do not start with cur.
+	// if no completion is found then fallback to default completion of shell.
+	`
+
+		while IFS='' read -r line; do COMPREPLY+=("$line"); done < <(compgen -o bashdefault -o default -o nospace -W "${opts}" -- "${cur}")
+
+		return 0
+	fi
+}
+
+# call the _s5cmd_cli_bash_autocomplete to complete s5cmd command. 
+complete -o nospace -F _s5cmd_cli_bash_autocomplete s5cmd
+`
+
+const pwsh = `$fn = $($MyInvocation.MyCommand.Name)
+$name = $fn -replace "(.*)\.ps1$", '$1'
+Register-ArgumentCompleter -Native -CommandName $name -ScriptBlock {
+	param($commandName, $wordToComplete, $cursorPosition)
+	$other = "$wordToComplete --generate-bash-completion"
+		Invoke-Expression $other | ForEach-Object {
+		[System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_)
+		}
+}
+`
+
+func getBashCompleteFn(cmd *cli.Command, isOnlyRemote, isOnlyBucket bool) func(ctx *cli.Context) {
+	isOnlyRemote = isOnlyRemote || isOnlyBucket
+	return func(ctx *cli.Context) {
+		var arg string
+		args := ctx.Args()
+		l := args.Len()
+
+		if l > 0 {
+			arg = args.Get(l - 1)
+		}
+
+		// argument may start with a quotation mark, in this case we want to trim that before
+		// checking if it has prefix s3://
+		// Beware that we only want to trim the first char, not all of the leading
+		// quotation marks, because those quotation marks may be actual charactes.
+		if strings.HasPrefix(arg, "'") {
+			arg = strings.TrimPrefix(arg, "'")
+		} else {
+			arg = strings.TrimPrefix(arg, "\"")
+		}
+
+		if isOnlyRemote || strings.HasPrefix(arg, "s3://") {
+			if strings.HasPrefix(arg, "-") {
+				cli.DefaultCompleteWithFlags(cmd)(ctx)
+				return
+			}
+			printS3Suggestions(ctx, arg, isOnlyBucket)
+			return
+		}
+
+		cli.DefaultCompleteWithFlags(cmd)(ctx)
+	}
+}
+
+// it returns a complete function which prints the argument, itself, which is to be completed.
+// If the argument is empty string it uses the defaultCompletions to make suggestions.
+func ineffectiveCompleteFnWithDefault(cmd *cli.Command, defaultCompletions ...string) func(ctx *cli.Context) {
```

Comment:
WDYM by ineffective? Please give it a name that indicates what it does.

Reply:
I mean it doesn't actually predict anything but merely returns the argument itself (or some default values) to override shell's default autocompletion. May be `nonpredictiveCompleteFnWithDefault` can be a better name (or worse?)

Reply:
For the record with https://github.com/Kucukaslan/s5cmd/commit/578f64c79aa9d76206c5fe90e8435a6afad91fdb this function is split to two parts and the auto-completion function is named `constantCompleteWithDefault`.